src/math/hough_transform_simd.cpp

```diff
@@ -0,0 +1,258 @@
+#include "hough_transform_simd.h"
+#include "../image_function_helper.h"
+#include "../penguinv/cpu_identification.h"
+
+#ifdef PENGUINV_AVX_SET
+#include <immintrin.h>
+#endif
+
+#ifdef PENGUINV_SSE_SET
+#include <emmintrin.h>
+#endif
+
+#ifdef PENGUINV_NEON_SET
+#include <arm_neon.h>
+#endif
+
+#include <algorithm>
+#include <cmath>
+
+namespace
+{
+    const bool float_layout_check  = std::is_standard_layout<PointBase2D<float>>::value;
+    const bool double_layout_check = std::is_standard_layout<PointBase2D<double>>::value;
+    const uint32_t avx_double = 4u;
+    const uint32_t avx_float = 8u;
+    const uint32_t sse_double = 2u;
+    const uint32_t sse_float = 4u;
+    const uint32_t neon_double = 2u;
+    const uint32_t neon_float = 4u;
+
+    const float minimumAngleStep = 0.001f * static_cast<float>( pvmath::pi ) / 180.0f;
+    const float minimumLineTolerance = 1e-5f;
+    
+    template <typename _Type>
+    void FindDistance( const std::vector< PointBase2D< _Type > > & input, std::vector < _Type > & distance, _Type cosVal, _Type sinVal, size_t offset = 0u )
+    {
+        _Type * distanceVal = distance.data();
+        const PointBase2D<_Type> * point = input.data() + offset;
+        const PointBase2D<_Type> * pointEnd = point + input.size();
+
+        for ( ; point != pointEnd; ++point, ++distanceVal )
+            (*distanceVal) = point->x * sinVal + point->y * cosVal;
+    }
+
+    template <typename _Type>
+    void FindDistanceSimd( const std::vector< PointBase2D< _Type > > & input, std::vector < _Type > & distance, _Type cosVal, _Type sinVal, const size_t inputPointCount )
+    {
+        FindDistance( input, distance, cosVal, sinVal );
+    }
+
+    template <>
+    void FindDistanceSimd< float >( const std::vector< PointBase2D< float > > & input, std::vector < float > & distance, float cosVal, float sinVal, const size_t inputPointCount )
+    {
+        simd::SIMDType simdType = simd::actualSimdType();
+
+        if ( simdType == simd::avx_function && float_layout_check ) {
+            #ifdef PENGUINV_AVX_SET
+            const size_t simdWidth = (inputPointCount*2) / (avx_float*2);
+            const size_t totalSimdWidth = simdWidth * (avx_float*2);
+            const size_t nonSimdWidth = (inputPointCount*2) - totalSimdWidth;
+
+            const float * point = reinterpret_cast<const float*>(input.data());
+            const float * PointEndSimd = point + totalSimdWidth;
+
+            float * distanceVal = distance.data();
+
+            const __m256 coeff = _mm256_set_ps(cosVal, sinVal, cosVal, sinVal, cosVal, sinVal, cosVal, sinVal);
+            const __m256i ctrl = _mm256_set_epi32(7, 6, 3, 2, 5, 4, 1, 0);
+
+            for( ;point != PointEndSimd; point += avx_float, distanceVal += avx_float )
+            {
+                __m256 src1 = _mm256_loadu_ps(point);
```

Comment:
I think you try to cast it to __m256 * and just use post increment (a++) here and for next load function. What do you think? No performance difference but little nicer code.